Add support for PostgreSQL arrays

vincentpoon opened the pull request:
This adds support for multidimensional ARRAY types in base-jdbc.  There's also an example implementation in PostgreSql.

ebyhr commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:
```diff
@@ -229,12 +234,36 @@ public JdbcTableHandle getTableHandle(SchemaTableName schemaTableName)
     }
 
+    // connectors should set the dimensions of any multidimensional ARRAY type, if supported
+    protected List<JdbcColumnHandle> setArrayDimensions(JdbcTableHandle tableHandle, List<JdbcColumnHandle> columns)
```
`getArrayDimensions` is better?

ebyhr commented on presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java:
```diff
@@ -84,6 +84,56 @@ public void testInsert()
+    public void testArrays()
+            throws Exception
+    {
+        assertUpdate("CREATE TABLE tmp_array1 AS SELECT ARRAY[1, 2, NULL] AS col", 1);
```
I would drop these tables.

ebyhr commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -112,6 +129,64 @@ protected ResultSet getTables(Connection connection, String schemaName, String t
+
+        if (arrayColumnDimensions.size() > 0) {
+            try (Connection connection = connectionFactory.openConnection()) {
+                String sql = String.format(
```
`String.format` → `format`

ebyhr commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -112,6 +129,64 @@ protected ResultSet getTables(Connection connection, String schemaName, String t
+                                "AND att.attname in ('%s')",
+                        tableHandle.getSchemaName(),
+                        tableHandle.getTableName(),
+                        arrayColumnDimensions.keySet().stream().collect(joining("','")));
```
How about 
```java
join("','", arrayColumnDimensions.keySet())
```

ebyhr commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -112,6 +129,64 @@ protected ResultSet getTables(Connection connection, String schemaName, String t
+        Map<String, Integer> arrayColumnDimensions = columns
+                .stream()
+                .filter(column -> column.getColumnType() instanceof ArrayType)
+                .collect(toMap(arrayColumn -> arrayColumn.getColumnName(), arrayColumn -> 0));
```
How about using method reference?
```java
.collect(toMap(JdbcColumnHandle::getColumnName, arrayColumn -> 0))
```

ebyhr commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -112,6 +129,64 @@ protected ResultSet getTables(Connection connection, String schemaName, String t
+                .filter(column -> column.getColumnType() instanceof ArrayType)
+                .collect(toMap(arrayColumn -> arrayColumn.getColumnName(), arrayColumn -> 0));
+
+        if (arrayColumnDimensions.size() > 0) {
```
How about ```!arrayColumnDimensions.isEmpty()```?

ebyhr commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -124,12 +199,62 @@ protected ResultSet getTables(Connection connection, String schemaName, String t
+    {
+        try (Connection connection = connectionFactory.openConnection()) {
+            // PostgreSql array typenames are the base element type prepended with "_"
+            String elementPgTypeName = arrayTypeHandle.getJdbcTypeName().replaceAll("_", "");
```
I feel `.substring(1)` is better than `replaceAll` if you want to just remove first character `_`.

vincentpoon commented:
Thanks @ebyhr , updated with your suggestions

findepi commented:
@ebyhr thanks for doing first review pass. Appreciated!

@vincentpoon  thanks for your PR!

In my opinion, there are some design changes to be done. In particular, I don't like the fact that we're constructing `List<JdbcColumnHandle>` and then remap them to another `List<JdbcColumnHandle>`. This is error prone, as it's possible to add future call site obtaining wrong `List<JdbcColumnHandle>` without remapping them. 

My suggestion regarding moving it forward:

1. Extract a commit introducing Block-typed read and write functions (call it eg `Add BlockReadFunction, BlockWriteFunction`).
   This change touches multiple files but otherwise is rather straightforward. You can even create a separate PR, so we can merge this sooner and have less code changes here.

2. Replace `private static ResultSet BaseJdbcClient#getColumns(JdbcTableHandle, java.sql.DatabaseMetaData)` with `protected Map<String,JdbcTypeHandle> getColumnTypes(ConnectorSession session, JdbcTableHandle)`. (map to be key-ed by column name)
    Default implementation would still use `java.sql.DatabaseMetaData#getColumns` of course.
    (Call this eg `Allow overriding querying column metadata`)

3. Add the code to handle postgres arrays.
    - extend `JdbcTypeHandle` with additional fields (not sure what these should be: "dimensions"? "elementType"?)
    - override this new `getColumnTypes` method in PostgresClient
    - ... i don't know yet how this should look like -- see below

I don't have yet answer on how  `arrayReadFunction` and `arrayWriteFunction` could look like.
What are the abstractions needed, what assumptions are OK to be made.
If we consider the fact that we don't a priori know how a base-jdbc -based connector is going to make in/out values, we probably don't know how it will map in/out array elements.
Maybe for starter, this should be a Postgres-specific feature?

cc @electrum

vincentpoon commented:
Thanks for looking at this @findepi .  Agree that the new methods in BaseJdbcClient here aren't ideal - I think that was a result of me trying to push as much reusable code up into BaseJdbcClient as possible.  This work actually came out of my work on the Phoenix connector (#76) as I figured I could get rid of a lot of array-specific stuff there if I made a more general solution.  But as you suggest, perhaps it's best to leave a lot of this in the specific jdbc implementations.

The `BlockReadFunction` and `BlockWriteFunction` are the only things I really need, i think.  I'll create a separate PR for that.

Let me try to rework this a little.  AFAIK, the main thing with jdbc ARRAY is there's no standard naming convention (but that's already true with other types, which is why we have e.g. WriteMapping#getDataType), and no standard in terms of what ResultSet#getArray returns you (as the javadoc clearly states).

vincentpoon commented:
@findepi I updated this, now that #454 is merged.  There's less impact on base-jdbc now, and all the ARRAY-related functionality is in presto-postgresql.

I couldn't quite implement `getColumnTypes()` the way you suggested, since there was a recent change that uses the `ResultSet` to get whether a column is nullable.  So in other words, `getColumns()` and `getJdbcTypeHandle()` both need to use the `ResultSet`, which is why I pass that as a parameter.

Let me know what you think, thanks!

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:
```diff
@@ -369,4 +399,87 @@ public static LongWriteFunction timestampWriteFunction()
+            return new Time((long) prestoNative);
+        }
+        else if (TIMESTAMP.equals(prestoType)) {
+            return new Timestamp((long) prestoNative);
```
That's not correct. With legacy timestamp semantics (default), `prestoNative` is a utc millis (Java timestamp) which interpreted in *session* zone yields year/month/day/hour/minutes that the value represents.
OTOH, `java.sql.TImestamp(x)` treats `x` as utc millis (Java timestamp) which interpreted in *JVM* zone yields year/month/day/hour/minutes that the value represents.
To fix this you would need:
- write different conversion code
- pass `ConnectorSession` so that the code can do condition on `io.prestosql.spi.connector.ConnectorSession#isLegacyTimestamp`

... and even then the code would not be generic. `java.sql.Timestamp` cannot represent certain values and many JDBC drivers (but not all) support `java.time.LocalDateTime` and we should use that.

For this reasons, the `prestoNativeToJdbcObject` cannot be done in a generic way, without connector-specific code. I would therefore suggest two things:

1. move this conversion to PostgresClient (or some companion helper utility class -- if possible)
   * then make sure there are no unnecessary changes in `StandardColumnMappings.java` (i would expect none are needed)
2. pass `ConnectorSession` here
3. remove support for time and timestamp types from this PR. It's easy to get them wrong and it's better that we focus on more generic stuff first and add timestamp arrays later.
   * to ensure we (and others) do not forget about timesamps and the necessary abstractions they require, I would keep the `TIMESTAMP.equals(prestoType)` condition like:
     ```
     if (TIMESTAMP.equals(prestoType)) {
         throw new PrestoException(.....); // TODO
     ```

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:
```diff
@@ -369,4 +399,87 @@ public static LongWriteFunction timestampWriteFunction()
+            }
+        }
+        else {
+            return prestoNative;
```
It's safer to throw when we don't know what to do:
```
throw new PrestoException( .... "Unsupported type: " + prestoType);
```

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:
```diff
@@ -369,4 +399,87 @@ public static LongWriteFunction timestampWriteFunction()
+                return ((Slice) prestoNative).getBytes();
+            }
+            else {
+                throw new PrestoException(FUNCTION_IMPLEMENTATION_ERROR, "Slice of unexpected type: " + prestoNative.getClass());
```
- `prestoType` is what actually matters, so include that in the message (`prestoNative.getClass()` will be always `Slice` here)
- i am not sure `FUNCTION_IMPLEMENTATION_ERROR` is the right error code... if you can't find anything more proper, use generic err code instead

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java:
```diff
@@ -369,4 +399,87 @@ public static LongWriteFunction timestampWriteFunction()
+        if (prestoNative == null) {
+            return null;
+        }
+        else if (prestoType instanceof DecimalType) {
```
nit: per the code style, we do not write `else` when not necessary, ie when the `then` block always returns.

I would remove `else` and also add blank space before subsequent `if` (here & for all cases below)

vincentpoon commented:
@findepi I've updated this with two separate commits.  The first only adds `arrayDimensions` to base-jdbc.  Everything else in the second commit is in presto-postgresql now.

Thanks for your helpful comments, it's looking a lot better now.

findepi commented:
> one more comment that i have now is about tests -- please look
> into how we usually test type mappings in `TestPostgreSqlTypeMapping`.

@vincentpoon to be clear: i am not sure this will be applicable. This needs trying / more thinking.

vincentpoon commented:
Thanks, I'll take a look at `TestPostreSqlTypeMapping`

vincentpoon commented:
@findepi I've pushed another commit with new `TestPostreSqlTypeMapping` tests.  They helped my catch a bug with certain `char[]` - I now throw `NOT_SUPPORTED` explicitly in `TypeUtils` to avoid this kind of error.

Also, apparently PostgreSQL's JDBC driver doesn't yet support varbinary arrays, so I removed support for that.

findepi commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/TypeUtils.java:
```diff
@@ -0,0 +1,292 @@
+        }
+
+        if (TIMESTAMP.equals(prestoType)) {
+            throw new PrestoException(NOT_SUPPORTED, "Unsupported type: " + prestoType); //TODO
```
I realized i contradicted myself (https://github.com/prestosql/presto/pull/317#discussion_r265917494 > last bullet).
So choose whichever you prefer. Sorry for this.

vincentpoon commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/PostgreSqlClient.java:
```diff
@@ -157,12 +255,63 @@ public WriteMapping toWriteMapping(ConnectorSession session, Type type)
             return WriteMapping.sliceMapping("bytea", varbinaryWriteFunction());
         }
+        if (TINYINT.equals(type)) {
+            return WriteMapping.longMapping("smallint", tinyintWriteFunction());
```
We can't test this with the framework in `TestPostgreSqlTypeMapping`, because Postgres doesn't actually have a `TINYINT` type.  It will get created here as a `SMALLINT` column in Postgres.  Thus when we read it back, we interpret it from the metadata as a `SMALLINT`, but `TestPostgreSqlTypeMapping` will expect a `TINYINT` type back.

Anyways, I will exclude this for now, and do it in another PR sometime.

findepi replied:
`TestPostgreSqlTypeMapping` has two data setups strategies:
 - `postgresCreateAndInsert` -- table is created in Postgres and populated in postgres
   - write is in Postgres
   - read is in Presto
- `prestoCreateAsSelect` -- table is created and populated in Presto (using CTAS)
   - write is in Presto
   - read is in Presto

The second strategy is the one that you want here, because the method you're fixing (`toWriteMapping`) is called only when Presto is writing.

vincentpoon replied:
It doesn't matter which you use, there is no `TINYINT` type in Postgres, so it will get created as a `SMALLINT` column in Postgres.  So with this change you can write `TINYINT` via Presto, but when `TestPostgreSqlTypeMapping` reads the data, it will fetch using the column metadata which is of `SMALLINT` type, but the test will expect `TINYINT`.  Thus, I cannot use `DataType.tinyintDataType()` directly.

I added a commit that tests by using a custom `DataType` - you can see whether it's acceptable.

vincentpoon commented on presto-postgresql/src/main/java/io/prestosql/plugin/postgresql/TypeUtils.java:
```diff
@@ -0,0 +1,292 @@
+            return utf8Slice((String) jdbcObject);
+        }
+
+        if (prestoType instanceof CharType) {
```
test is in table `prestoCreateAsSelect("test_array_parameterized_char_unicode")`

vincentpoon commented on presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java:
```diff
@@ -231,6 +235,128 @@ public void testDecimalExceedingPrecisionMax()
     }
 
+    @Test
+    public void testArray()
```
I removed `TestPostgreSqlIntegrationSmokeTest#testArrays`.
I think `TestPostgreSqlTypeMapping` is good enough

vincentpoon commented:
@findepi Thanks again for the review, I have pushed a new squashed revision.  I believe I have addressed all the comments, and have replied to some specifically.

findepi commented:
LGTM. Thanks @vincentpoon for working on this.
Can you please rebase?

vincentpoon commented:
@findepi  rebased

vincentpoon commented:
Thanks @findepi , addressed your latest comments

findepi commented:
Merged, thanks!

findepi commented:
Note: as per https://github.com/prestosql/presto/pull/687, this needs to be explicitly enabled.

cc @hashbash (https://github.com/prestodb/presto/issues/11422), @troels (https://github.com/prestodb/presto/pull/8849)
